File: examples/atb_models/tests/pythontest/operations_test/linear_parallel/test_linear_parallel.py
# Copyright (c) Huawei Technologies Co., Ltd. 2024-2025. All rights reserved.
# MindIE is licensed under Mulan PSL v2.
# You can use this software according to the terms and conditions of the Mulan PSL v2.
# You may obtain a copy of Mulan PSL v2 at:
#          http://license.coscl.org.cn/MulanPSL2
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
# EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
# MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
# See the Mulan PSL v2 for more details.
import random
import unittest
import json
import logging

import torch
import torch_npu
import torch.multiprocessing as mp

from operations_test import operation_test


def linear_parallel_worker(rank, world_size, world_input):
    torch_npu.npu.set_device(rank)
    m, k, n, x, weight = world_input

    # setUp
    op_type = "LinearParallel"
    op_param = {
        "rank": rank,
        "rankSize": world_size,
        "rankRoot": 0,
    }
    op_name = "LinearParallelOperation"

    # input for rowparallel
    slice_start, slice_end = k // world_size * rank, k // world_size * (rank + 1)
    if rank == world_size - 1:
        slice_end = k    

    in_tensor_0 = x[:, slice_start: slice_end]
    in_tensor_1 = weight[:, slice_start: slice_end]
    out_tensor_0 = torch.zeros(m, n, dtype=torch.float16)
    inputs = {'in0': in_tensor_0.npu(), 'in1': in_tensor_1.npu()}
    outputs = {'out0': out_tensor_0.npu()}

    # forward
    operation = operation_test.atb.BaseOperation(op_type=op_type,
                                   op_param=json.dumps(op_param),
                                        op_name=op_name)
    operation.forward(inputs, outputs)
    
    # gloden
    golden_out_tensor = torch.matmul(x.npu(), weight.npu().T)
    out_tensor = list(outputs.values())[0]

    if not golden_compare(out_tensor.cpu(), golden_out_tensor.cpu()):
        raise ValueError("Output tensor does not match the golden tensor.")


def golden_compare(out_tensor, golden_out_tensor, rtol=0.0001, atol=0.0001):
    result = torch.allclose(out_tensor, golden_out_tensor, rtol=rtol, atol=atol)
    if not result:
        logger = logging.getLogger()
        logger.info("out_tensor.shape: %s, golden_out_tensor.shape: %s", out_tensor.shape, golden_out_tensor.shape)
        logger.info("out_tensor:\n %s", out_tensor)
        logger.info("golden_out_tensor:\n %s", golden_out_tensor)
    return result


class TestLinearParallelOperation(operation_test.OperationTest):
    def setUp(self):
        self.world_size = 2        
        self.m = random.randint(1, 100)
        self.k = random.randint(1, 100)
        self.n = random.randint(1, 100)

    def test_float16(self):
        world_input_x = torch.rand(self.m, self.k, dtype=torch.float16)
        world_input_weight = torch.rand(self.n, self.k, dtype=torch.float16)
        world_input = (self.m, self.k, self.n, world_input_x, world_input_weight)

        mp.spawn(linear_parallel_worker, nprocs=self.world_size, args=(self.world_size, world_input))
    
    
if __name__ == '__main__':
    unittest.main()

File: examples/atb_models/tests/fuzztest/core/test_linear_parallel.cpp
/**
 * Copyright (c) Huawei Technologies Co., Ltd. 2024. All rights reserved.
 * MindIE is licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *          http://license.coscl.org.cn/MulanPSL2
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
 * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
 * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 */

#include <gtest/gtest.h>
#include <atb/atb_infer.h>
#include "atb_speed/log.h"
#include "operations/fusion/linear/linear_parallel.h"
#include "../utils/fuzz_utils.h"
#include "secodeFuzz.h"

namespace atb_speed {
TEST(LinearParallelDTFuzz, LinearParallel)
{
    std::srand(time(NULL));
    ATB_SPEED_LOG_DEBUG("begin====================");
    std::string fuzzName = "LinearParallelDTFuzzLinearParallel";
    DT_FUZZ_START(0, 10000, const_cast<char*>(fuzzName.c_str()), 0) {
        uint32_t fuzzIndex = 0;
        atb_speed::common::LinearParallelParam param;

        atb_speed::common::FusionLinearParam fusionLinearParam;
        int LinearQuantTypeEnumTable[] = {0, 1, 2, 3, 4, 5};
        int TransposeTypeEnumTable[] = {-1, 0, 1};
        fusionLinearParam.quantType = static_cast<atb_speed::common::LinearQuantType>(
            *(unsigned int *) DT_SetGetNumberEnum(
                &g_Element[fuzzIndex++], 0, LinearQuantTypeEnumTable, 6));
        fusionLinearParam.isBF16 = FuzzUtil::GetRandomBool(fuzzIndex);
        fusionLinearParam.hasBias = FuzzUtil::GetRandomBool(fuzzIndex);
        fusionLinearParam.transposeType = static_cast<atb_speed::common::TransposeType>(
            *(int *) DT_SetGetNumberEnum(
                &g_Element[fuzzIndex++], -1, TransposeTypeEnumTable, 3));
        param.fusionLinearParam = fusionLinearParam;

        int LinearParallelTypeEnumTable[] = {0, 1, 2};
        param.parallelType = static_cast<atb_speed::common::LinearParallelType>(
            *(unsigned int *) DT_SetGetNumberEnum(
                &g_Element[fuzzIndex++], 0, LinearParallelTypeEnumTable, 3));
        param.biasAfterSync = FuzzUtil::GetRandomBool(fuzzIndex);
        param.unpadInputs = FuzzUtil::GetRandomBool(fuzzIndex);
        param.supportLcoc = FuzzUtil::GetRandomBool(fuzzIndex);

        atb_speed::common::TensorParallelInfo tensorParallelInfo;
        int worldSize = *(int *) DT_SetGetNumberRange(&g_Element[fuzzIndex++], 1, 1, 8);
        tensorParallelInfo.rank = *(int *) DT_SetGetNumberRange(&g_Element[fuzzIndex++], 0, 0, worldSize);
        tensorParallelInfo.worldSize = worldSize;
        std::vector<std::string> backendEnumTable = {"lccl", "hccl"};
        tensorParallelInfo.backend = backendEnumTable[*(int *) DT_SetGetNumberRange( \
            &g_Element[fuzzIndex++], 0, 0, 1)];
        param.tensorParallelInfo = tensorParallelInfo;

        atb::Node linearParallelNode;
        LinearParallel(param, &linearParallelNode.operation);
    }
    DT_FUZZ_END()
    SUCCEED();
}
}